internal/slicer/report.go

```diff
@@ -0,0 +1,64 @@
+package slicer
+
+import (
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	"github.com/canonical/chisel/internal/fsutil"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+type ReportEntry struct {
+	Path   string
+	Mode   fs.FileMode
+	Hash   string
+	Size   int
+	Slices map[*setup.Slice]bool
+	Link   string
+}
+
+// Report holds the information about files and directories created when slicing
+// packages.
+type Report struct {
+	// Root is the filesystem path where the all reported content is based.
+	Root string
+	// Entries holds all reported content, indexed by their path.
+	Entries map[string]ReportEntry
+}
+
+// NewReport returns an empty report for content that will be based at the
+// provided root path.
+func NewReport(root string) *Report {
+	return &Report{
+		Root:    root,
+		Entries: make(map[string]ReportEntry),
+	}
+}
+
+func (r *Report) Add(slice *setup.Slice, info *fsutil.Info) error {
+	if !strings.HasPrefix(info.Path, r.Root) {
+		return fmt.Errorf("internal error: cannot add path %q outside out root %q", info.Path, r.Root)
```

Comment:
s/outside out/outside of/

This is also not an internal error. Internal errors are those that are impossible to get to unless the given type/function/whatever was coded incorrectly. Calling this function with wrong parameters is trivial to do without any mistakes in the code of the Report.

The error message might also be just a tad more clear, as it's not mentioning the issue is related to the report.

("cannot report path %q when the report root is at %q", Path, Root)

Reply:
Right, what I wanted to convey with `internal error` is that we should not return this error to the user. Because the user has not done anything wrong but rather there is an error in Chisel itself, I wanted to signal that it is a bug on our side. How could we do that?

Reply:
Discussed this in the meeting, I agree to drop the "internal error" prefix. A more user friendly solution would need custom error types and better reporting, let's wait until the issue comes up.